src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,904 @@
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        /* set to store ip address ranges we should masquerade for */
+        r = nft_new_set(nfnl, &batch[msgcnt], family, NFT_SYSTEMD_MASQ_SET_NAME, ++set_id, NFT_SET_INTERVAL, ip_type, ip_type_size);
+        if (r < 0)
+                goto out_unref;
+
+        /*
+         * map to store ip address:port pair to dnat to.  elements in concatenation
+         * are rounded up to 4 bytes.
+         *
+         * Example: ip protocol . tcp daddr is sizeof(uint32_t) + sizeof(uint32_t), not
+         * sizeof(uint8_t) + sizeof(uint16_t).
+         */
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        r = nft_new_map(nfnl, &batch[msgcnt], family, NFT_SYSTEMD_DNAT_MAP_NAME, ++set_id, 0,
+                        concat_types2(TYPE_INET_PROTOCOL, TYPE_INET_SERVICE), sizeof(uint32_t) * 2,
+                        concat_types2(ip_type, TYPE_INET_SERVICE), ip_type_size + sizeof(uint32_t));
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        r = sd_nfnl_message_new_dnat_rule_pre(nfnl, &batch[msgcnt], family, "prerouting");
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        r = sd_nfnl_message_new_dnat_rule_out(nfnl, &batch[msgcnt], family, "output");
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        r = sd_nfnl_message_new_masq_rule(nfnl, &batch[msgcnt], family, "postrouting");
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        r = sd_nfnl_message_batch_end(nfnl, &batch[msgcnt]);
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt <= NFT_INIT_MSGS);
+        r = nfnl_netlink_sendv(nfnl, batch, msgcnt);
+        if (r == -EEXIST)
+                r = 0;
+
+out_unref:
+        for (i = 0; i < msgcnt; i++)
+                sd_netlink_message_unref(batch[i]);
+
+        return r;
+}
+
+int fw_nftables_init(FirewallContext *ctx) {
+        _cleanup_(sd_netlink_unrefp) sd_netlink *nfnl = NULL;
+        int r;
+
+        r = sd_nfnl_socket_open(&nfnl);
+        if (r < 0)
+                return r;
+
+        r = fw_nftables_init_family(nfnl, AF_INET);
+        if (r < 0)
+                return r;
+
+        ctx->nfnl = TAKE_PTR(nfnl);
+        return 0;
+}
+
+void fw_nftables_exit(FirewallContext *ctx) {
+        ctx->nfnl = sd_netlink_unref(ctx->nfnl);
+}
+
+static int nft_message_add_setelem_iprange(sd_netlink_message *m,
+                                           const union in_addr_union *source,
+                                           unsigned int prefixlen) {
+        uint32_t mask, start, end;
+        unsigned int nplen;
+        int r;
+
+        assert(prefixlen <= 32);
+        nplen = 32 - prefixlen;
+
+        mask = (1U << nplen) - 1U;
```

Comment:
if npen is allowed to be 32 (i.e. prefixlen is allowed to be zero), then this will result in undefined behaviour, since 1U << 32 is not defined, and will overflow.

if prefixlen is now allowed to be zero, then maybe error out early?

(it looks to me as if a full ip range should be ok to support here though, hence maybe special case prefixlen == 0 here and allow it?)

Reply:
Its not allowed anymore, this will yield -EINVAL earlier.
The reason is two-fold:
1. The caller is networkd, and afaics it always passes in the netmask configured on the interface.
2. The existing iptables backend (which the nft one is modeled on) doesn't restrict based on interface. IOW, when using /0, it will masquerade everything on any interface, which is probably not wanted.  If this should be supported, it might be best to add extra special code that just deletes the existing rule (which checks if a given address is in a range we're masquerading for) and replaces that with a single masquerade rule with no further checks.